cvmfs/bundle_mgr.cc

```diff
@@ -8,47 +8,131 @@
-    is_valid_ &= cvmfs::GetDirentForInode(mp, mp->file_system(), ino, &dirent_);
-    fname_ = GetFileName(path_);
-    parent_path_ = GetParentPath(path_);
-
-    // There is a naming convention regarding the name of the file with the
-    // contents of the bundle
-    bundle_file_path_ = PathString(parent_path_.ToString() + "/.cvmfsbundle."
-                                   + fname_.ToString());
-
-    bfm_ = new BundleFileMgr(bundle_file_path_);
+#include "fetch.h"
+#include "util/inode.h"
+#include "util/posix.h"
+
+BundleMgr::BundleMgr(MountPoint *mp, fuse_ino_t ino) : mount_point_(mp) {
+  is_valid_ = cvmfs::GetPathForInode(mp, mp->file_system(), ino, &path_);
+  is_valid_ &= cvmfs::GetDirentForInode(mp, mp->file_system(), ino, &dirent_);
+  if (not is_valid_) {
+    return;
   }
+  fname_ = GetFileName(path_);
+  parent_path_ = GetParentPath(path_);
+  fetcher_ = dirent_.IsExternalFile() ? mp->external_fetcher() : mp->fetcher();
 
-void BundleMgr::Fetch() {
-  std::vector<pthread_t> fetcher_pool;
+  // There is a naming convention regarding the name of the file with the
+  // contents of the bundle
+  bundle_file_path_ = PathString(parent_path_.ToString() + "/.cvmfsbundle."
+                                 + fname_.ToString());
 
-  for (PathString next_f = bfm_->GetNext(); not next_f.IsEmpty();
-       next_f = bfm_->GetNext()) {
-    pthread_t fetcher_t;
+  bfm_ = new BundleFileMgr(bundle_file_path_);
+  pipe_bm_[0] = pipe_bm_[1] = -1;
+}
 
-    // TODO(christge): DoFetch requires real arguments
-    pthread_create(&fetcher_t, nullptr, DoFetch, nullptr);
-    fetcher_pool.emplace_back(fetcher_t);
+void BundleMgr::Fetch() {
+  SpawnFetchers();
+
+  CacheManager::LabeledObject *obj;
+  while ((obj = bfm_->GetNext()) != nullptr) {
+    // TODO(christge): In this form the assignment of a fetcher is left to
+    // SendLabeledObject. This is incorrect. SendLabeledObject should only
+    // unpack the LabeledObject and send it over the pipe
+    SendLabeledObject(*obj);
   }
 
-  // Join fetcher threads
-  while (not fetcher_pool.empty()) {
-    for (auto it = fetcher_pool.begin(); it != fetcher_pool.end();) {
-      auto &thread = *it;
-      void *retval;
-      int is_joinable = pthread_tryjoin_np(thread, &retval);
-      if (is_joinable == 0) {
-        // fetch has finished
-        fetcher_pool.erase(it);
-      } else {
-        // still busy
-        ++it;
+  JoinFetchers();
+}
+
+void BundleMgr::JoinFetchers() {
+  bool detach_mode = false;
+  // Join fetcher threads.
+  // Give a fetcher thread 10 seconds to join.
+  for (auto it = fetcher_pool_.begin(); it != fetcher_pool_.end(); ++it) {
+    auto &tuple = *it;
+    auto thread = std::get<0>(tuple);
+    auto fd = std::get<1>(tuple);
+
+    struct timespec ts;
+    if (clock_gettime(CLOCK_REALTIME, &ts) == -1) {
+      LogCvmfs(kLogBungleMgr,
+               kLogDebug,
+               "Failed to read CLOCK_REALTIME. Detaching fetchers.");
+      detach_mode = true;
```

Comment:
This is hard to test. I think we should just expect to get a timestamp and otherwise crash.